haskell/repl.bzl

```diff
@@ -0,0 +1,571 @@
+    if not pattern_str.startswith("//"):
+        fail("Invalid haskell_repl pattern. Patterns must start with //.")
+
+    # Separate package and target (if present).
+    package_target = pattern_str[2:].split(":", maxsplit = 2)
+    package_str = package_target[0]
+    target_str = None
+    if len(package_target) == 2:
+        target_str = package_target[1]
+
+    # Parse package pattern.
+    package = []
+    dotdotdot = False  # ... has to be last component in the pattern.
+    for s in package_str.split("/"):
+        if dotdotdot:
+            fail("Invalid haskell_repl pattern. ... has to appear at the end.")
+        if s == "...":
+            dotdotdot = True
+        package.append(s)
+
+    # Parse target pattern.
+    if dotdotdot:
+        if target_str != None:
+            fail("Invalid haskell_repl pattern. ... has to appear at the end.")
+    elif target_str == None:
+        if len(package) > 0 and package[-1] != "":
+            target_str = package[-1]
+        else:
+            fail("Invalid haskell_repl pattern. The empty string is not a valid target.")
+
+    return struct(
+        package = package,
+        target = target_str,
+    )
+
+def _match_label(pattern, label):
+    """Whether the given local workspace label matches any of the patterns.
+
+    Args:
+      patterns: A list of parsed patterns to match the label against.
+        Apply `_parse_pattern` before passing patterns into this function.
+      label: Match this label against the patterns.
+
+    Returns:
+      A boolean. True if the label is in the local workspace and matches any of
+      the given patterns. False otherwise.
+
+    """
+
+    # Only local workspace labels can match.
+    # Despite the docs saying otherwise, labels don't have a workspace_name
+    # attribute. So, we use the workspace_root. If it's empty, the target is in
+    # the local workspace. Otherwise, it's an external target.
+    if label.workspace_root != "":
+        return False
+
+    package = label.package.split("/")
+    target = label.name
+
+    # Match package components.
+    for i in range(min(len(pattern.package), len(package))):
+        if pattern.package[i] == "...":
+            return True
+        elif pattern.package[i] != package[i]:
+            return False
+
+    # If no wild-card or mismatch was encountered, the lengths must match.
+    # Otherwise, the label's package is not covered.
+    if len(pattern.package) != len(package):
+        return False
+
+    # Match target.
+    if pattern.target == "all" or pattern.target == "*":
+        return True
+    else:
+        return pattern.target == target
+
+def _load_as_source(include, exclude, lbl):
+    """Whether a package should be loaded by source or as binary."""
+    for pat in exclude:
+        if _match_label(pat, lbl):
+            return False
+
+    for pat in include:
+        if _match_label(pat, lbl):
+            return True
+
+    return False
```

Comment:
> If I remember correctly, there is a way to call bazel query from within bazel rules, via some builtin primitive. If its possible, we should use that at least.

I'm not aware of any primitive or Starlark library for this task. The only related feature I'm aware of is [`genquery`](https://docs.bazel.build/versions/master/be/general.html#genquery). However, we cannot use `genquery` for two reasons. First, `genquery` does not allow wildcards:

> The only difference between the queries allowed here and on the command line is that queries containing wildcard target specifications (e.g. //pkg:* or //pkg:all) are not allowed here. 

Second, `genquery` uses `bazel query` instead of `bazel cquery`, so you quickly encounter issues like the following:

```
build aborted: no such target '@nixpkgs_cc_toolchain//:bin/clang': target 'bin/clang'
```

>  If not, we should open an upstream feature request describing our use-case and mark the attributes with an experimental- prefix for now (until we know whether there is going to be an upstream feature).

If opened a feature request on the Bazel tracker: https://github.com/bazelbuild/bazel/issues/7763